snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -151,3 +165,72 @@ def terminate(self) -> None:
     def join(self, timeout: Optional[float] = None) -> None:
         self.__next_step.close()
         self.__next_step.join(timeout)
+
+
+class ScheduleSubscriptions(ProcessingStrategy[CommittableTick]):
+    """"
+    This strategy is responsible for scheduling all of the subscriptions
+    for a given tick.
+
+    The subscriptions to be scheduled are those that:
+    - Have the same partition index as the tick received, and
+    - Are scheduled to be run within the time interval defined by the tick
+
+    For backward compatibility, this assumes that the number of partitions of
+    the subscription storage matches the number of partitions of the main topic
+    being subscribed to.
+
+    The subscriptions to be scheduled are are encoded and produced to the
+    scheduled topic to be picked up by the subscription executor to run
+    the actual query later.
+
+    If all of the scheduled subscription messages for the tick are successfully
+    produced, and tick specifies `should_commit` = True, then the offset is committed.
+    """
+
+    def __init__(
+        self,
+        entity_key: EntityKey,
+        schedulers: Mapping[int, Scheduler[Subscription]],
+        producer: Producer[KafkaPayload],
+        scheduled_topic_spec: KafkaTopicSpec,
+        commit: Callable[[Mapping[Partition, Position]], None],
+    ) -> None:
+        self.__schedulers = schedulers
+        self.__encoder = SubscriptionScheduledTaskEncoder(entity_key)
+        self.__producer = producer
+        self.__scheduled_topic = Topic(scheduled_topic_spec.topic_name)
+        self.__commit = commit
+        self.__closed = False
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[CommittableTick]) -> None:
+        assert not self.__closed
+
+        tick = message.payload.tick
+
+        tasks = self.__schedulers[tick.partition].find(tick.timestamps)
+        futures = [
+            self.__producer.produce(self.__scheduled_topic, self.__encoder.encode(task))
+            for task in tasks
+        ]
+
+        for future in as_completed(futures):
+            future.result()
```

Comment:
I don't think you need to wait for this at this point.
I would at least wait for the callbacks only when you know you are about to commit. So you can keep processing and producing till a checkpoint where you need to be sure all messages have been sent before going on.

You can even go on eagerly processing further messages till you are done with the callbacks for the batch before committing:
- produce all the messages for the ticks until you get a `should_commit` = `true`
- At that point, if all futures have completed (likely not), keep processing and keep the commit offset aside
- While you process new messages keep checking if all callbacks for the first batch have completed. 
- When they are commit and continue.

Reply:
Adding some context. This does not require any major redesign:
See the StreamProcessor class from arroyo and the ProcessingStrategy docstring.
What the StreamProcessor does is:
- poll from the consumer
- poll on the processing strategy 
- submit to the processing strategy
- repeat

The rationale behind that is:
- all these methods are asynchronous (they return nothing), which means potentially none of them needs to wait on anything. This does not need to have any batching step where we stop for a specified amount of time every specified amount of time. We could be able to continue processing elements one by one. Each step can tell if it is falling behind by raising `MessageRejected`
- When you call submit (the processor calls submit), that is supposed to be an immediate operation, it schedules work to do, it is not required to do anything, it should not make the caller wait. It can also exert back-pressure by raising MessageRejected if there are too many items (messages) to be processed.
- At every step poll is called to actually trigger work. Poll should not wait either but it may be impractical to avoid.

This actually can help you in avoiding the wait and optimizing the time you spend executing.
In theory in order to follow the abstraction you could do it this way:
- submit stores the tick in a list and does nothing else. Before that it checks if you have pending messages you are waiting the callback from. If that list is too long raise MessageRejected to exert backpressure.
- poll will do what this method actually does except for waiting. Though it checks which futures are completed (which is not an operation you wait onto). It does not wait. As soon as it finds a committable tick to be completed it calls commit. 
- The code here never waits on anything. It is the StreamProcessor that interprets `MessageRejected` as "please wait".

I don't think we ever used this system that way (which is how it is supposed to be used), but this seems a valid use case. In practice it may cause a tight loop as we never wait on the future and keep iterating trying to submitting, that would increase CPU usage, though I think it is worth trying, we can always add a waiting operation on the futures in the poll method to avoid that.

Reply:
Thanks, this was very helpful advice. I rewrote like you suggested to make both poll() and submit() non blocking and raise the MessageRejected to slow down the consumer.